### PR TITLE
Add WatchTask requirement for TaskLauncher

### PR DESCRIPTION
**Pull request by mitchellh**
This adds a new API `WatchTask` that must be implemented by `TaskLauncher` plugins.

`WatchTask` receives the state result from `StartTask` and is expected to block while it is running. WatchTask must return a `TaskResult` struct which currently only contains an exit code. Additionally, `WatchTask` has access to a ✨ UI ✨ object so it can also stream output. It is expected to stream platform output if it has it (i.e. pod launching information).

This API _may or may not_ be called and task launchers shouldn't depend on it. It is a nice-to-have that is sometimes called by Waypoint core in certain occasions to get more information from tasks.

### CODE REVIEW

File: component/task.go [hunk 1]
```diff
@@ -22,3 +22,9 @@ type TaskLaunchInfo struct {
 	// arguments to that program.
 	Arguments []string
 }
+
+// TaskResult is the result value that must be returned by the RunTask
+// function.
+type TaskResult struct {
+	ExitCode int
+}
```
Comment (briancain) [inline component/task.go:30]: While we're here, is there an easy way to place the spawned Task ids from the plugin outside of the Tasks plugin proto result into its own variable? Not on this struct `TaskResult`, but probably somewhere here inside `component/task`? If it's a bigger change, feel free to ignore for now.

File: internal/plugin/task.go [hunk 2]
```diff
@@ -265,6 +303,49 @@ func (s *taskLauncherServer) StopTask(
...
+		argmapper.Typed(internal),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	ret := &pb.TaskWatch_Resp{}
+	if r, ok := result.(*component.TaskResult); ok {
+		ret.ExitCode = int32(r.ExitCode)
+	}
```
Comment (briancain) [inline internal/plugin/task.go:344]: Is there anything we can do in the `!ok` case? Should we just return an error that the implemented plugin was not properly implemented to return a `TaskResult`? Maybe we would of already caught that before now in the compiler.

File: internal/plugin/task_test.go [hunk 3]
```diff
@@ -59,6 +59,48 @@ func TestTaskLauncherStart(t *testing.T) {
...
+	mockB.On("WatchTaskFunc").Return(watchFunc)
+
+	plugins := Plugins(WithComponents(mockB), WithMappers(testDefaultMappers(t)...))
+	client, server := plugin.TestPluginGRPCConn(t, plugins[1])
+	defer client.Close()
+	defer server.Stop()
+
+	raw, err := client.Dispense("tasklauncher")
+	require.NoError(err)
+	fmt.Printf("=> %T\n", raw)
```
Comment (briancain) [inline internal/plugin/task_test.go:84]: Debug print?

**Review by briancain — APPROVED** [review]
Looks good! ⌚ ✅

**Review by evanphx — APPROVED** [review]
Looks good, I think the question I have is elsewhere, namely where do we flow the UI events to...

**Comment by mitchellh** [comment]
> Looks good, I think the question I have is elsewhere, namely where do we flow the UI events to...

This is called by a runner, and the runner passes in the UI, so it goes to the job the runner is running. But, the plugin itself doesn't concern itself with that since in theory the UI could go anywhere.

**Pull request merged**